File: operators/grpc_operators/cpp/server/grpc_server_request.cpp
/*
 * SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights
 * reserved. SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "grpc_server_request.hpp"

namespace holoscan::ops {

void GrpcServerRequestOp::setup(OperatorSpec& spec) {
  spec.param(request_queue_, "request_queue", "Request Queue", "Incoming gRPC requests.");
  spec.param(allocator_,
             "allocator",
             "Allocator",
             "Allocator used for converting incoming EntityRequest objects to a GXF Entity.");
  spec.param(rpc_timeout_,
             "rpc_timeout",
             "RPC Call timeout",
             "Timeout in seconds for the gRPC server to issue a Finish command if no data is"
             "is transmitted or received.");

  spec.output<holoscan::gxf::Entity>("output");

  cuda_stream_handler_.define_params(spec);
}

void GrpcServerRequestOp::compute(InputContext& op_input, OutputContext& op_output,
                                  ExecutionContext& context) {
  if (!request_queue_->empty()) {
    auto request = request_queue_->pop();
    auto result = nvidia::gxf::Entity(std::move(*request));
    op_output.emit(result, "output");
  }
}

std::shared_ptr<UnboundedAllocator> GrpcServerRequestOp::allocator() {
  return allocator_.get();
}

const uint32_t GrpcServerRequestOp::rpc_timeout() {
  return rpc_timeout_.get();
}

const cudaStream_t GrpcServerRequestOp::get_cuda_stream() {
  return cuda_stream_handler_.get_cuda_stream(fragment()->executor().context());
}

}  // namespace holoscan::ops

File: operators/grpc_operators/python/server/grpc_server_request.py
# SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import logging

from holoscan.core import Fragment, Operator, OperatorSpec
from holoscan.resources import Allocator, CudaStreamPool, UnboundedAllocator

from operators.grpc_operators.python.common.conditional_variable_queue import ConditionVariableQueue
from operators.grpc_operators.python.common.tensor_proto import TensorProto


class GrpcServerRequestOp(Operator):
    def __init__(
        self,
        fragment: Fragment,
        request_queue: ConditionVariableQueue,
        allocator: Allocator,
        cuda_stream_pool: CudaStreamPool,
        rpc_timeout: int = 5000,
        *args,
        **kwargs,
    ):
        self.request_queue: ConditionVariableQueue = request_queue
        self._allocator: Allocator = allocator
        self._cuda_stream_pool: CudaStreamPool = cuda_stream_pool
        self._rpc_timeout: int = rpc_timeout
        self.logger: logging.Logger = logging.getLogger(__name__)

        if not isinstance(request_queue, ConditionVariableQueue):
            raise ValueError("request_queue must be a ConditionVariableQueue")

        super().__init__(fragment, *args, **kwargs)

    @property
    def allocator(self) -> UnboundedAllocator:
        return self._allocator

    @property
    def rpc_timeout(self) -> int:
        return self._rpc_timeout

    @property
    def cuda_stream_pool(self) -> CudaStreamPool:
        return self._cuda_stream_pool

    def setup(self, spec: OperatorSpec):
        spec.output("output")

    def compute(self, op_input, op_output, context):
        if not self.request_queue.empty():
            request = self.request_queue.pop()

            entity = TensorProto.proto_to_tensor(request, context)
            op_output.emit(entity, "output")
